src/constraints/shake_gpu.jl

```diff
@@ -0,0 +1,807 @@
+# The RATTLE Equations are modified from LAMMPS:
+# https://github.com/lammps/lammps/blob/develop/src/RIGID/fix_rattle.cpp
+
+
+# Returns a if condition is true, b otherwise (without branching)
+@inline function branchless_select(condition, a, b)
+    return condition * a + (!condition) * b
+end
+
+@inline function branchless_min(a, b)
+    return branchless_select(a <= b, a, b)
+end
+
+
+@inline function solve2x2exactly(λ, A, C)
+
+    determinant = A[1, 1] * A[2, 2] - A[1, 2] * A[2, 1]
+
+    if iszero(determinant)
+        error("SHAKE determinant is zero, cannot solve")
+    end
+
+    inv_det = 1.0 / determinant
```

Comment:
`1.0` is `Float64` and will promote the type, use `inv` or `one(T)`. Might speed things up on GPU for `Float32`. Also in other functions.